torch/csrc/jit/passes/device_type_analysis.cpp

```diff
@@ -0,0 +1,237 @@
+#include <ATen/core/interned_strings.h>
+#include <ATen/core/jit_type.h>
+#include <c10/core/Device.h>
+#include <c10/util/ArrayRef.h>
+#include <c10/util/Optional.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/device_type_analysis.h>
+#include <torch/library.h>
+#include <memory>
+#include "ATen/core/List.h"
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+using Tensor = at::Tensor;
+using Device = at::Device;
+
+using PropRule = std::function<bool(Node*)>;
+/*
+A Propagation Rule takes the Node, and
+applies the relevant properties to the Tensor outputs
+of the Node (based on the rule itself)
+
+Returns: Bool indicating if anything was changed
+*/
+
+bool setDeviceType(Value* value, Device device, bool can_overwrite = false) {
+  auto tensor_type = value->type()->cast<TensorType>();
+  TORCH_INTERNAL_ASSERT(tensor_type, "Expecting a tensor type");
+  if (!tensor_type->device().has_value()) {
+    value->setType(tensor_type->withDevice(device));
+    return true;
+  }
+  if (tensor_type->device().value() != device) {
+    TORCH_INTERNAL_ASSERT(
+        can_overwrite,
+        "Expected device to be ",
+        device,
+        " but found ",
+        tensor_type->device().value());
+    value->setType(tensor_type->withDevice(device));
+    return true;
+  }
+  return false;
+}
+
+bool setReturnsToDevice(Node* n, Device device) {
+  bool changed = false;
+  for (Value* out : n->outputs()) {
+    auto tensor_type = out->type()->cast<TensorType>();
+    if (!tensor_type) {
+      continue;
+    }
+    changed |= setDeviceType(out, device);
+  }
+  return changed;
+}
+
+bool propWithNoDevice(Node* n) {
+  // Figure out what the common device to propagate is
+  c10::optional<Device> device;
+  for (Value* inp : n->inputs()) {
+    auto tensor_type = inp->type()->cast<TensorType>();
+    if (!tensor_type) {
+      continue;
+    }
+    if (device) {
+      if (tensor_type->device().has_value()) {
+        auto cur_device = tensor_type->device().value();
+        TORCH_INTERNAL_ASSERT(
+            device == cur_device,
+            "Expected device to be ",
+            device.value(),
+            " but found ",
+            cur_device);
+      }
+      continue;
+    } else {
+      device = tensor_type->device();
+    }
+  }
+
+  return device.has_value() && setReturnsToDevice(n, device.value());
```

Comment:
To confirm: Because Torchscript is SSA, if there is a device that is already set, it must have been set by something other than Device Propagation, right?

Is the concern that some other pass would have set an incorrect type?

Reply:
Yea, it's basically that the dtypes in the graph should reflect what we can currently infer. As an edge case, imagine that we had a constant tensor of dim 0 that made it so we were correctly inferring cpu as an output of an broadcast node if the other was a cpu tensor but failed if the other was a cuda tensor. if someone updated the input dtype to cuda, we shouldn't have the prior, no longer relevant dtype present on the node